File: solutions/haskell/083-reorder-list.hs
{-
  Problem 83: Reorder List (LeetCode 143)
  Difficulty: Med
  Language: Haskell
-}
reorderList :: [a] -> [a]
reorderList xs =
  let n = length xs
      mid = n `div` 2
      first = take mid xs
      second = reverse (drop mid xs)
      merge [] bs = bs
      merge as [] = as
      merge (a:as') (b:bs') = a : b : merge as' bs'
  in merge first second

File: solutions/clojure/083-reorder-list.clj
;; Problem 83: Reorder List (LeetCode 143)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn reorder-list [lst]
  (let [mid (quot (count lst) 2)
        first (take mid lst)
        second (reverse (drop mid lst))]
    (letfn [(merge [a b]
              (cond (empty? a) b
                    (empty? b) a
                    :else (concat [(first a) (first b)]
                                  (merge (rest a) (rest b)))))]
      (merge first second))))
